react/ProductCustomizer.js

```diff
@@ -1,60 +1,317 @@
+  * Call optional variations parser and calculates the total.
+  * @param object variationObject
+  * @return void
+  */
+  handleVariationChange = async variationObject => {
+    await this.handleSelectedVariation(variationObject)
+    this.handleParseOptionalVariationsBySkuId(variationObject.skuId)
+    this.calculateTotalFromSelectedVariation()
+  }
+
+  /**
+  * handleParseOptionalVariationsBySkuId
+  * Create the index state of numeric stepper.
+  * @param string skuId
+  * @return void
+  */
+  handleParseOptionalVariationsBySkuId = skuId => {
+    const {
+      productQuery: {
+        product,
+      },
+    } = this.props
+
+    const sku = product.items.find(sku => {
+      return sku.itemId === skuId
+    })
+    const optionalVariations = this.parseAttachments('optionals', sku)
+
+    this.createNumericStepperIndexesStates(optionalVariations.variations)
+    this.setState({ optionalVariations })
+  }
+
+  /**
+  * createNumericStepperIndexesStates
+  * Create the initial state of Numeric Stepper Component.
+  * @param array items
+  * @return void
+  */
+  createNumericStepperIndexesStates = items => {
+    const choosedAmount = {}
+
+    items.forEach(item => {
+      choosedAmount[item.name] = 0
+    })
+
+    this.setState({ choosedAmount })
+  }
+
+  /**
+  * onHandleNumericStepperChange
+  * Sets the optional variation values by index.
+  * @param object variationObject
+  * @return void
+  */
+  onHandleNumericStepperChange = variationObject => {
+    const {
+      choosedAmount,
+    } = this.state
+
+    choosedAmount[variationObject.index] = variationObject.quantity
+
+    this.setState({ choosedAmount })
   }
 
+  /**
+  * handleSelectedVariation
+  * Resets extra variations state and update the current variation selected.
+  * @param object variationObject
+  * @return void
+  */
+  handleSelectedVariation = variationObject => {
+    this.setState({ extraVariations: [] })
+
+    return this.setState({
+      selectedVariation: {
+        skuId: variationObject.skuId,
+        variation: variationObject.variation,
+        quantity: variationObject.quantity,
+      },
+    })
+  }
+
+  /**
+  * handleSelectedExtraVariations
+  * Add the changed optional variation and call the calculate method
+  * @param object variationObject
+  * @return void
+  */
+  handleSelectedExtraVariations = async variationObject => {
+    const currentExtraVariations = this.state.extraVariations
+    const key = currentExtraVariations.findIndex(extraVariation => {
+      return extraVariation.index === variationObject.index
+    })
+
+    this.onHandleNumericStepperChange(variationObject)
+
+    if (key === -1) {
+      currentExtraVariations.push(variationObject)
+    } else {
+      if (variationObject.quantity !== 0) {
+        currentExtraVariations[key] = variationObject
+      } else {
+        currentExtraVariations.splice(key)
+      }
+    }
+
+    await this.setState({ extraVariations: currentExtraVariations })
```

Comment:
Could you explain what is the meaning of this `await` statement? The `this.setState` do not return a promise (actually, it returns nothing, since it is a setter function) and as you can see [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#Description) this will be converted to a simple resolved promise.

Reply:
I'll remove these await statement and gonna put the method below into a setState's callback